### PR TITLE
Plan b fix

### PR DESCRIPTION
**Pull request by jmillan**

### CODE REVIEW

File: src/handlers/sdp/MediaSection.ts [hunk 1]
```diff
@@ -691,6 +691,52 @@ export class OfferMediaSection extends MediaSection
 			? encoding.rtx.ssrc
 			: undefined;
 
+		for (const codec of offerRtpParameters!.codecs)
```
Comment (ibc) [inline src/handlers/sdp/MediaSection.ts]:
```suggestion
		for (const codec of offerRtpParameters.codecs)
```

`!` not needed right?

File: src/handlers/sdp/MediaSection.ts [hunk 2]
```diff
@@ -691,6 +691,52 @@ export class OfferMediaSection extends MediaSection
...
+		{
+			const rtp: any =
+				{
+					payload : codec.payloadType,
+					codec   : getCodecName(codec),
+					rate    : codec.clockRate
+				};
```
Comment (ibc) [inline src/handlers/sdp/MediaSection.ts]:
```suggestion
			const rtp: any =
			{
				payload : codec.payloadType,
				codec   : getCodecName(codec),
				rate    : codec.clockRate
			};
```

Pretty sure we use this syntax everywhere :)

File: src/handlers/sdp/MediaSection.ts [hunk 3]
```diff
@@ -691,6 +691,52 @@ export class OfferMediaSection extends MediaSection
...
+			if (codec.channels! > 1)
+				rtp.encoding = codec.channels;
+
+			this._mediaObject.rtp.push(rtp);
+
+			const fmtp =
+				{
+					payload : codec.payloadType,
+					config  : ''
+				};
```
Comment (ibc) [inline src/handlers/sdp/MediaSection.ts]:
```suggestion
			const fmtp =
			{
				payload : codec.payloadType,
				config  : ''
			};
```

File: src/handlers/sdp/MediaSection.ts [hunk 4]
```diff
@@ -691,6 +691,52 @@ export class OfferMediaSection extends MediaSection
...
+						payload : codec.payloadType,
+						type    : fb.type,
+						subtype : fb.parameter
+					});
+			}
+		}
+
+		this._mediaObject.payloads += ' '+ offerRtpParameters!.codecs
+		.map((codec: RtpCodecParameters) => codec.payloadType)
+		.join(' ');
```
Comment (ibc) [inline src/handlers/sdp/MediaSection.ts]:
```suggestion
		this._mediaObject.payloads += ' ' + offerRtpParameters
			.codecs
			.map((codec: RtpCodecParameters) => codec.payloadType)
			.join(' ');
```

File: src/handlers/sdp/MediaSection.ts [hunk 5]
```diff
@@ -745,6 +791,18 @@ export class OfferMediaSection extends MediaSection
...
+
+		this._mediaObject.payloads = this._mediaObject.payloads.split(' ')
+			.filter((payload: any) => !payloads.includes(Number(payload)))
+			.join(' ');
+
+		this._mediaObject.rtp = this._mediaObject.rtp.filter((rtp: any) => !payloads.includes(rtp.payload));
+
+		this._mediaObject.rtcpFb = this._mediaObject.rtcpFb.filter((rtcpFb: any) => !payloads.includes(rtcpFb.payload));
+
+		this._mediaObject.fmtp = this._mediaObject.fmtp.filter((fmtp: any) => !payloads.includes(fmtp.payload));
```
Comment (ibc) [inline src/handlers/sdp/MediaSection.ts]: Why those additions? I need some rationale here.

File: test/test.js [hunk 6]
```diff
@@ -1517,6 +1519,78 @@ test('connection state change does not fire "connectionstatechange" in closed Tr
 	sendTransport.removeAllListeners('connectionstatechange');
 }, 500);
 
+test('RemoteSdp properly handles multiple streams of the same type in planB', async () =>
+{
+	let remoteSdp = new RemoteSdp({ planB: true });
+	let sdp = undefined;
+	let sdpObject = undefined;
+
+	await remoteSdp.receive({
```
Comment (ibc) [inline test/test.js]: Is this passing ESLint? `{` should be in next line.

File: test/test.js [hunk 7]
```diff
@@ -1517,6 +1519,78 @@ test('connection state change does not fire "connectionstatechange" in closed Tr
...
+	expect(sdpObject.media[0].rtp[1].payload).toBe(102);
+	expect(sdpObject.media[0].rtp[1].codec).toBe('rtx');
+	expect(sdpObject.media[0].rtp[2].payload).toBe(103);
+	expect(sdpObject.media[0].rtp[2].codec).toBe('H264');
+	expect(sdpObject.media[0].rtp[3].payload).toBe(104);
+	expect(sdpObject.media[0].rtp[3].codec).toBe('rtx');
+	expect(sdpObject.media[0].ssrcs.length).toBe(8);
+
+
+	await remoteSdp.planBStopReceiving({
```
Comment (ibc) [inline test/test.js]: And this.

**Comment by jmillan** [comment]
running lint..

File: test/test.js (hunk 6 above)
Comment (jmillan) [reply test/test.js]: yes, it was passing. Updated.

File: src/handlers/sdp/MediaSection.ts (hunk 5 above)
Comment (jmillan) [reply src/handlers/sdp/MediaSection.ts]: I'm removing from the SDP the description of the payloads being stopped. This way, if those same payloads are sent again (imagine webcam stop/start), the corresponding description of the new payloads will be added again, representing exactly what is being received at each time.

I'm looking for the standard way to do this, but PlanB is not even a standard. Any comment on how we should be doing this?

Comment (ibc) [reply src/handlers/sdp/MediaSection.ts]: understood

**Review by ibc — APPROVED** [review]
Tested?

**Comment by jmillan** [comment]
> Tested?

Tested with ReactNative handler. Works like a charm.

**Comment by jmillan** [comment]
Let me try to test is with chrome 67.

**Comment by nazar-pc** [comment]
Chrome 67 is from the first half of 2018, who cares (about Chrome 67 and Plan B in general)?

**Comment by ibc** [comment]
> Chrome 67 is from the first half of 2018, who cares (about Chrome 67 and Plan B in general)?

react-native-webrtc

**Comment by jmillan** [comment]
react-native-webrtc uses M89 at the moment

**Comment by jmillan** [comment]
> react-native-webrtc uses M89 at the moment

But planB, this is tested.

**Comment by nazar-pc** [comment]
Oh, they still didn't manage to migrate to Unified plan :disappointed: 
> With plan B going away soon, this PR fill find its way in before the end of the year.

Sorry for off-topic.

**Comment by jmillan** [comment]
Just a note.

Before this PR, endpoints using planB couldn't negotiate streams with different codecs for the same media type. Whether this will work for older Chromes I did not test, but neither it worked before.

**Comment by jmillan** [comment]
I've tested Chrome60. I can't find any 67-70 version. And the renegotiation for different codecs for same media type does not work neither, even though the SDP looks file. Note that this did not work before this PR neither.

Should we go ahead with this PR anyway? It adds a missing functinality to PlanB with react-native-webrtc, solely, as it seems.

**Comment by jmillan** [comment]
> And the renegotiation for different codecs for same media type does not work neither, even though the SDP looks file.

No errors, I did not check further. Wondering if it's worth it the time for trying to make this old Chrome versions to support this **new** feature.

**Comment by ibc** [comment]
Do you mean that this PR:

- Fixes an issue when in react-native-webrtc.
- Does not fix same issue with Chrome 60.
- It does not break anything in Chrome 60.

?

If so, IMHO go ahead.

**Comment by jmillan** [comment]
> Fixes an issue when in react-native-webrtc.
> Does not fix same issue with Chrome 60.
> It does not break anything in Chrome 60.

Yes, this is what I see

**Review by ibc — APPROVED** [review]

**Pull request merged**